velox/functions/prestosql/ArraysOverlap.cpp

```diff
@@ -0,0 +1,160 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "velox/functions/lib/LambdaFunctionUtil.h"
+
+namespace facebook::velox::functions {
+namespace {
+
+template <typename T>
+class ArraysOverlapFunction : public exec::VectorFunction {
+ public:
+  ArraysOverlapFunction() {}
+
+  explicit ArraysOverlapFunction(
+      SetWithNull<T> constantSet,
+      bool isLeftConstant)
+      : constantSet_(std::move(constantSet)), isLeftConstant_(isLeftConstant) {}
+
+  void apply(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args,
+      const TypePtr& /* outputType */,
+      exec::EvalCtx* context,
+      VectorPtr* result) const override {
+    // if one of them is a ConstantVector.
+    // Iterate through the other ArrayVector->elementsFlatVector
+    // and do constantSet_.find()
+    BaseVector* left = args[0].get();
+    BaseVector* right = args[1].get();
+    if (constantSet_.has_value() && isLeftConstant_) {
+      std::swap(left, right);
+    }
+    exec::LocalDecodedVector decoder(context, *left, rows);
+    exec::LocalDecodedVector elementsDecoder(context);
+    auto decodedLeftElements =
+        getDecodedElementsFromArrayVector(decoder, elementsDecoder, rows);
+
+    FlatVector<bool>* resBoolVector = (*result)->asFlatVector<bool>();
+    auto processRow = [&](const vector_size_t rowId,
+                          const SetWithNull<T>& rightSet) {
+      auto decodedLeftArray = decoder.get();
+      auto baseLeftArray = decodedLeftArray->base()->as<ArrayVector>();
```

Comment:
These 2 lines can be moved out of processRow. The local variables of the function can be accessed in the lambda as local variables.

Reply:
I moved extracting the decodedLeftArray and baseLeftArray as part of getDecodedElementsFromArrayVector.
That is the reason I need to re-fetch them here.